iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/metadata/fetcher/cache/TableDeviceSchemaCache.java

```diff
@@ -84,39 +95,147 @@ public Map<String, String> getDeviceAttribute(
     }
   }
 
-  public void update(
+  public void updateAttributes(
       final String database,
       final String tableName,
       final String[] deviceId,
       final Map<String, String> attributeMap) {
+    dualKeyCache.update(
+        new TableId(database, tableName),
+        new TableDeviceId(deviceId),
+        new TableDeviceCacheEntry(),
+        entry -> entry.updateAttribute(database, tableName, attributeMap),
+        false);
+  }
+
+  // Shall pass in "null" for deviceId when invalidating attribute for a table
+  public void invalidateAttributes(
+      final String database, final String tableName, final String[] deviceId) {
+    dualKeyCache.update(
+        new TableId(database, tableName),
+        Objects.nonNull(deviceId) ? new TableDeviceId(deviceId) : null,
+        new TableDeviceCacheEntry(),
+        entry -> -entry.invalidateAttribute(),
+        false);
+  }
+
+  /////////////////////////////// Last Cache ///////////////////////////////
+
+  // The input "TimeValuePair" shall never contain null value
+  public void updateLastCache(
+      final String database,
+      final String tableName,
+      final String[] deviceId,
+      final Map<String, TimeValuePair> measurementUpdateMap) {
     readWriteLock.readLock().lock();
     try {
-      final TableDeviceCacheEntry entry =
-          dualKeyCache.get(new TableId(database, tableName), new TableDeviceId(deviceId));
-      if (Objects.nonNull(entry)) {
-        entry.update(attributeMap);
-      }
+      forceUpdateCache(database, tableName, deviceId, measurementUpdateMap);
     } finally {
       readWriteLock.readLock().unlock();
     }
   }
 
+  // The input "TimeValuePair" shall never contain null value
+  public void mayUpdateLastCacheWithoutLock(
+      final String database,
+      final String tableName,
+      final String[] deviceId,
+      final Map<String, TimeValuePair> measurementUpdateMap) {
+    if (putLastCacheWhenWriting) {
+      forceUpdateCache(database, tableName, deviceId, measurementUpdateMap);
+    } else {
+      dualKeyCache.update(
+          new TableId(database, tableName),
+          new TableDeviceId(deviceId),
+          new TableDeviceCacheEntry(),
+          entry -> entry.tryUpdate(database, tableName, measurementUpdateMap),
+          false);
+    }
+  }
+
+  private void forceUpdateCache(
+      final String database,
+      final String tableName,
+      final String[] deviceId,
+      final Map<String, TimeValuePair> measurementUpdateMap) {
+    dualKeyCache.update(
+        new TableId(database, tableName),
+        new TableDeviceId(deviceId),
+        new TableDeviceCacheEntry(),
+        entry -> entry.updateLastCache(database, tableName, measurementUpdateMap),
+        true);
+  }
+
+  public TimeValuePair getLastEntry(
+      final String database,
+      final String tableName,
+      final String[] deviceId,
+      final String measurement) {
+    final TableDeviceCacheEntry entry =
```

Comment:
add java doc for all these public methods in TableDeviceSchemaCache, like whether database contains `root.` and deviceId contains database.tableName for its first index, and can measurement be `time`?